### PR TITLE
Feature/edit registration

### PR DESCRIPTION
**Pull request by PaulDoyle-DEFRA**
This PR includes the full front-office edit/deregistration journey, including entering email address and registration reference, generating an email with an edit magic link, the main edit form, integration with existing edit pages, declaration and confirmation pages.
It also includes a substantial refactoring of back-office edit functionality to avoid name clashes with the new front-office edit functionality.
This branch includes functionality from  [feature/RUBY-2566-wex-front-office-changes-user-journey](https://github.com/DEFRA/waste-exemptions-engine/tree/feature/RUBY-2566-wex-front-office-changes-user-journey) and makes PR https://github.com/DEFRA/waste-exemptions-engine/pull/837 redundant.
https://eaflood.atlassian.net/browse/RUBY-2566
https://eaflood.atlassian.net/browse/RUBY-2563
https://eaflood.atlassian.net/browse/RUBY-2561
https://eaflood.atlassian.net/browse/RUBY-2562
https://eaflood.atlassian.net/browse/RUBY-2598
https://eaflood.atlassian.net/browse/RUBY-2602
https://eaflood.atlassian.net/browse/RUBY-2565
https://eaflood.atlassian.net/browse/RUBY-2643
https://eaflood.atlassian.net/browse/RUBY-2644

### CODE REVIEW

File: spec/requests/waste_exemptions_engine/front_office_edit_declaration_forms_spec.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+module WasteExemptionsEngine
+  RSpec.describe FrontOfficeEditDeclarationFormsController do
```
Comment (jjromeo) [inline spec/requests/waste_exemptions_engine/front_office_edit_declaration_forms_spec.rb:6]: Should this have a string describe block like the other request specs?

File: spec/requests/waste_exemptions_engine/front_office_edit_forms_spec.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            expect(response.body).to include edit_exemptions_front_office_edit_forms_path(transient_registration_token)
+            expect(response.body).to include contact_name_front_office_edit_forms_path(transient_registration_token)
+            expect(response.body).to include contact_phone_front_office_edit_forms_path(transient_registration_token)
+            expect(response.body).to include contact_email_front_office_edit_forms_path(transient_registration_token)
+          end
+        end
+      end
+    end
+
+    describe "POST front_office_edit_form" do
```
Comment (jjromeo) [inline spec/requests/waste_exemptions_engine/front_office_edit_forms_spec.rb:66]: These tests are simply to clarify the user is redirected to the same page regardless of changes being made?

File: spec/models/waste_exemptions_engine/registration_spec.rb [hunk 3]
```diff
@@ -247,6 +245,38 @@ module WasteExemptionsEngine
...
+        let(:registration) { create(:registration) }
+
+        it { expect { regenerate_token }.to change(registration, :edit_token) }
+
+        it { expect { regenerate_token }.to change(registration, :edit_token_created_at) }
+
+        it do
+          regenerate_token
+
+          expect(registration.edit_token_created_at).to be_within(3).of(Time.zone.now)
```
[the remark this answers is not in the document]
Comment (PaulDoyle-DEFRA) [reply spec/models/waste_exemptions_engine/registration_spec.rb:265]: `3.seconds` - fixed

File: app/views/waste_exemptions_engine/front_office_edit_forms/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      <% transient_registration = @front_office_edit_form.transient_registration %>
+
+      <dl class="govuk-summary-list">
+
+        <div class="govuk-summary-list__row">
+          <dt class="govuk-summary-list__key">
+            <%= t(".row_labels.exemptions") %>
+          </dt>
+          <dd class="govuk-summary-list__value">
+            <%= transient_registration.exemptions.pluck(:code).join(",") %>
```
[the remark this answers is not in the document]
Comment (PaulDoyle-DEFRA) [reply app/views/waste_exemptions_engine/front_office_edit_forms/new.html.erb]: Agree - fixed

File: app/models/waste_exemptions_engine/front_office_edit_registration.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      return true if exemptions.pluck(:id).sort != registration.exemptions.pluck(:id).sort
+
+      %i[
+        contact_first_name
+        contact_last_name
+        contact_phone
+        contact_email
+      ].each do |attr|
+        return true if send(attr) != registration.send(attr)
+      end
```
[the remark this answers is not in the document]
Comment (PaulDoyle-DEFRA) [reply app/models/waste_exemptions_engine/front_office_edit_registration.rb]: Nice

File: spec/requests/waste_exemptions_engine/front_office_edit_declaration_forms_spec.rb (hunk 1 above)
Comment (PaulDoyle-DEFRA) [reply spec/requests/waste_exemptions_engine/front_office_edit_declaration_forms_spec.rb:6]: Yes, the request spec shouldn't assume anything about implementing classes - fixed

File: spec/requests/waste_exemptions_engine/front_office_edit_forms_spec.rb (hunk 2 above)
Comment (PaulDoyle-DEFRA) [reply spec/requests/waste_exemptions_engine/front_office_edit_forms_spec.rb:66]: Yes, this one is a bit atypical in that it needs to redirect to the same form regardless of user choice.

File: config/locales/forms/capture_complete_forms/en.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+en:
+  waste_exemptions_engine:
+    capture_complete_forms:
+      new:
+        title: If the email address you entered is correct we will send a link to that email address
+        heading: If the email address you entered is correct we will send a link to that email address
+        description: You can use the links in the email to update your contact details and deregister any or all of your exemptions. The link will expire after 48 hours.
```
Comment (brujeo) [inline config/locales/forms/capture_complete_forms/en.yml:7]: probably newline is missing here

**Review by brujeo — DISMISSED** [review]

**Review by brujeo — APPROVED** [review]

**Pull request merged**